python/example/berry/mcmc.py

```diff
@@ -0,0 +1,98 @@
+# flake8: noqa E402
+import os
+
+os.environ["XLA_FLAGS"] = "--xla_force_host_platform_device_count=6"
+import jax
+import jax.numpy as jnp
+import numpy as np
+import numpyro
+import numpyro.distributions as dist
+import numpyro.infer
+
+# This line is critical for enabling 64-bit floats.
+from jax.config import config
+
+config.update("jax_enable_x64", True)
+
+
+@profile
+def mcmc_berry(data, logit_p1, suc_thresh, n_arms=4, dtype=np.float64, n_samples=10000):
+    @profile
+    def mcmc_berry_model(y, n):
+        mu = numpyro.sample("mu", dist.Normal(-1.34, 10))
+
+        # An attempt at sampling from the logarithm to make the conditioning better.
+        # LogTransform = dist.transforms._InverseTransform(dist.transforms.ExpTransform())
+        # log_inverse_gamma = dist.TransformedDistribution(
+        #     dist.InverseGamma(0.0005, 0.000005),
+        #     LogTransform
+        # )
+        # log_sigma2 = numpyro.sample("log_sigma2", log_inverse_gamma)
+        # sigma2 = numpyro.deterministic("sigma2", jnp.exp(log_sigma2))
+
+        sigma2 = numpyro.sample("sigma2", dist.InverseGamma(0.0005, 0.000005))
+        with numpyro.plate("j", n_arms):
+            theta = numpyro.sample(
+                "theta",
+                dist.Normal(mu, jax.numpy.sqrt(sigma2)),
+            )
+            numpyro.sample(
+                "y",
+                dist.BinomialLogits(theta + logit_p1, total_count=n),
+                obs=y,
+            )
+
+    # Number of devices to pmap over
+    n_data = data.shape[0]
+    rng_keys = jax.random.split(jax.random.PRNGKey(0), n_data)
+
+    # NOTE: pmap requires an exact match between data.shape[0] and the number of
+    # devices. so to use it further with more than n_cores datasets, we would
+    # need to batch the data and layer a vmap underneath
+    # traces = jax.pmap(do_mcmc)(rng_keys, data[..., 0], data[..., 1])
```

Comment:
Jax documents how to use pmap with vmap, which I shamelessly ripped off here
https://github.com/pyro-ppl/numpyro/blob/6d82a6bec66c8e080a4f258ef7c9f4cab9d1ac2f/numpyro/infer/mcmc.py#L237

Reply:
I think that's not quite what we want here. That is doing:
1. a pmap over multiple datasets
2. a vmap (implicitly) over the chains underneath. 

What we want is instead:
1. a pmap over each chunk of datasets
2. a vmap over each dataset in the chunks.

This actually has me realizing that, because the parallelization overhead will be small compared to the fairly long runtime of the MCMC code, that the better way to do this is to actually flip it around.
1. Write a python loop that chunks the datasets into blocks of n_cores.
2. pmap over n_cores datasets. 

I think this third version is what we want.

Reply:
I'm going to do this because running the MCMC code is painful sometimes and having it more parallelized will be very nice.

Reply:
Yeah, I think the most-useful part of this snippet is just showing how to combine multiple results. More inspiring than directly useful.

Reply:
Great!

Reply:
I wrote a new parallel version that handles arbitrary numbers of dataset. It's chunking up the data nicely and is much faster than the non-parallel version.